numba/_lapack.c

```diff
@@ -0,0 +1,1035 @@
+ * to a Fortran integer.
+ *
+ * Parameters:
+ * kind the kind of val
+ * val  a pointer to the value to cast
+ *
+ * Returns:
+ * A Fortran int from a cast of val (in complex case, takes the real part).
+ *
+ */
+// a template would be nice
+// struct access via non c99 (python only) cmplx types, used for compatibility
+F_INT cast_from_X(char kind, void *val)
+{
+    switch(kind)
+    {
+        case 's':
+            return (F_INT)(*((float *) val));
+        case 'd':
+            return (F_INT)(*((double *) val));
+        case 'c':
+            return (F_INT)(*((npy_complex64 *)val)).real;
+        case 'z':
+            return (F_INT)(*((npy_complex128 *)val)).real;
+        default:
+        {
+            PyGILState_STATE st = PyGILState_Ensure();
+            PyErr_SetString(PyExc_ValueError,\
+                            "invalid kind in cast");
+            PyGILState_Release(st);
+        }
+    }
+    return -1;
+}
+
+// real space eigen systems info from dgeev/sgeev
+NUMBA_EXPORT_FUNC(F_INT)
+numba_raw_rgeev(char kind, char jobvl, char jobvr,
+                F_INT n, void *a, F_INT lda, void *wr, void *wi, void *vl,
+                F_INT ldvl, void *vr, F_INT ldvr, void *work, F_INT lwork,
+                F_INT * info)
+{
+    void *raw_func = NULL;
+
+    switch (kind)
+    {
+        case 's':
+            raw_func = get_clapack_sgeev();
+            break;
+        case 'd':
+            raw_func = get_clapack_dgeev();
+            break;
+        default:
+        {
+            PyGILState_STATE st = PyGILState_Ensure();
+            PyErr_SetString(PyExc_ValueError,\
+                            "invalid kind of real space *geev call");
+            PyGILState_Release(st);
+        }
+        return -1;
+    }
+    if (raw_func == NULL)
+        return -1;
+
+    (*(rgeev_t) raw_func)(&jobvl, &jobvr, &n, a, &lda, wr, wi, vl, &ldvl, vr,
+                          &ldvr, work, &lwork, info);
+    return 0;
+}
+
+// real space eigen systems info from dgeev/sgeev
+// as numba_raw_rgeev but the allocation and error handling is done for the user
+// Args are as per LAPACK.
+NUMBA_EXPORT_FUNC(F_INT)
+numba_ez_rgeev(char kind, char jobvl, char jobvr,
+               F_INT n, void *a, F_INT lda, void *wr, void *wi, void *vl,
+               F_INT ldvl, void *vr, F_INT ldvr)
+{
+    F_INT info = 0;
+    size_t base_size = -1;
+    // find the function to call, decide on a base type size
+    switch (kind)
+    {
+        case 's':
+            base_size = sizeof(float);
+            break;
+        case 'd':
+            base_size = sizeof(double);
+            break;
+        default:
+        {
+            PyGILState_STATE st = PyGILState_Ensure();
+            PyErr_SetString(PyExc_ValueError,\
+                            "Invalid kind in numba_ez_rgeev");
+            PyGILState_Release(st);
+        }
+        return -1;
+    }
+    F_INT lwork = -1;
+    void * work = malloc(base_size);
```

Comment:
Not sure why you are heap-allocating a temporary float or double. Can't you just use a variable on the stack?